grid_map_ros/include/grid_map_ros/Costmap2DConverter.hpp

```diff
@@ -0,0 +1,314 @@
+   * itself. To avoid introducing error, you want to shift the center so the
+   * underlying costmap2d cells align with the newly created map object. This does
+   * that here.
+   *
+   * @warning this does not lock the costmap2d object, you should take care to do so from
+   * outside this function in scope with any addLayerFromCostmap2DAtRobotPose calls you wish to make.
+   *
+   * @param costmap2d : the underlying Costmap2DROS object
+   * @param geometry: size of the subwindow to snapshot around the robot pose
+   * @param outputMap : initialise the output map with the required parameters
+   */
+  bool initializeFromCostmap2DAtRobotPose(costmap_2d::Costmap2DROS& costmap2d, const Length& length,
+                                          MapType& outputMap)
+  {
+    const double resolution = costmap2d.getCostmap()->getResolution();
+
+    // Get the Robot Pose Transform.
+    tf::Stamped<tf::Pose> tfPose;
+    if(!costmap2d.getRobotPose(tfPose))
+    {
+      std::ostringstream error_message;
+      error_message << "Could not get robot pose, is it actually published?";
+      ROS_ERROR_STREAM("CostmapConverter:: " << error_message.str());
+      return false;
+      // throw std::runtime_error(error_message.str());
+    }
+
+    // Determine new costmap origin.
+    Position robotPosition(tfPose.getOrigin().x() , tfPose.getOrigin().y());
+    Position rosMapOrigin(costmap2d.getCostmap()->getOriginX(), costmap2d.getCostmap()->getOriginY());
+    Position newCostMapOrigin;
+
+    // Note:
+    //   You cannot directly use the robot pose as the new 'costmap center'
+    //   since the underlying grid is not necessarily exactly aligned with
+    //   that (two cases to consider, rolling window and globally fixed).
+    //
+    // Relevant diagrams:
+    //  - https://github.com/ethz-asl/grid_map
+
+    // Float versions of the cell co-ordinates, use static_cast<int> to get the indices.
+    Position robotCellPosition = (robotPosition - rosMapOrigin) / resolution;
+
+    // if there is an odd number of cells
+    //   centre of the new grid map in the centre of the current cell
+    // if there is an even number of cells
+    //   centre of the new grid map at the closest vertex between cells
+    // of the current cell
+    int number_of_cells_x = length.x()/resolution;
+    int number_of_cells_y = length.y()/resolution;
+    if ( number_of_cells_x % 2 ) { // odd
+      newCostMapOrigin(0) = std::floor(robotCellPosition.x())*resolution + resolution/2.0 + rosMapOrigin.x();
+    } else {
+      newCostMapOrigin(0) = std::round(robotCellPosition.x())*resolution + rosMapOrigin.x();
+    }
+    if ( number_of_cells_y % 2 ) { // odd
+      newCostMapOrigin(1) = std::floor(robotCellPosition.y())*resolution + resolution/2.0 + rosMapOrigin.y();
+    } else {
+      newCostMapOrigin(1) = std::round(robotCellPosition.y())*resolution + rosMapOrigin.y();
+    }
+
```

Comment:
Aye, thought it should be simple as well. One of the other lads here shifted to index first, and then back to position, but had problems with numerical accuracies when sitting on the boundary. Shifting back and forth seems like overkill to me. Those functions don't entirely suit either - what we are seeking for here is the `mapPosition` variable, which in both of those methods is a known incoming variable of the method.

Not sure also it would be eminently reusable. We're seeking to align a grid map on top of an 'other' map that is as logically close as possible to a specified position. This needs information about the 'other' map (e.g. what is it's convention for the origin?). I'd leave this here and if we find another use case emerging, then do something about it.